### PR TITLE
Fix a number of compiler warnings and enable -Werror in Travis

### PR DESCRIPTION
**Pull request by simonduq**
FYI this is what a Werror build looks like (with all fixes from this PR included)
https://travis-ci.org/simonduq/contiki/builds/83121859

### CODE REVIEW

**Comment by simonduq** [comment]
(sdcc does not seem to understand `-Werror`)

**Comment by simonduq** [comment]
Now updated with more fixes from @adamdunkels, and with -Werror enabled in Travis (for most platforms)

**Comment by adamdunkels** [comment]
This pull request (finally) makes Contiki build cleanly with no warnings out of the box, for most platforms. There are still a number of warnings left for some platforms, which in most cases reflect real problems with the code (e.g. the AVR plaforms have a too small `clock_time_t` that causes problems with the RPL code).

This lets us catch problems early from now on, as every new pull request will have to build without warnings to get a green travis.

:+1: from me. Would be great if someone else could look through the changes too, as there are many of them.

File: apps/servreg-hack/servreg-hack.c [hunk 1]
```diff
@@ -332,12 +332,12 @@ static void
...
 {
   int numregs;
-  int flags;
+  /*int flags;*/
   int i;
   int bufptr;
 
   numregs = buf[MSG_NUMREGS_OFFSET];
-  flags   = buf[MSG_FLAGS_OFFSET];
+  /*flags   = buf[MSG_FLAGS_OFFSET];*/
```
Comment (oliverschmidt) [inline apps/servreg-hack/servreg-hack.c]: I dislike commented-out code without any hint why it is commented-out.

File: apps/webbrowser/www.c [hunk 2]
```diff
@@ -864,6 +864,7 @@ htmlparser_newline(void)
 
   wptr = webpageptr - WWW_CONF_WEBPAGE_WIDTH;
   petsciiconv_topetscii(wptr, WWW_CONF_WEBPAGE_WIDTH);
+  (void)wptr;
```
Comment (oliverschmidt) [inline apps/webbrowser/www.c]: I suggest to rather put the original two lines inside an `#ifdef WITH_PETSCII`.

File: core/ctk/ctk-filedialog.c [hunk 3]
```diff
@@ -155,14 +155,14 @@ ctk_filedialog_eventhandler(struct ctk_filedialog_state *s,
...
-      if((ctk_arch_key_t)data == CH_CURS_UP) {
+      if((unsigned long)data == CH_CURS_UP) {
 	clearptr();
 	if(fileptr > 0) {
 	  --fileptr;
 	}
 	showptr();
 	return 1;
-      } else if((ctk_arch_key_t)data == CH_CURS_DOWN) {
+      } else if((unsigned long)data == CH_CURS_DOWN) {
```
Comment (oliverschmidt) [inline core/ctk/ctk-filedialog.c]: See above.

File: cpu/avr/bootloader.c [hunk 4]
```diff
@@ -12,7 +12,7 @@
 /* MCUSR is a deprecated name but older avr-libc versions may define it */
 #if !defined (MCUCSR)
 # if defined (MCUSR)
-#  warning *** MCUCSR not defined, using MCUSR instead ***
+/*#  warning *** MCUCSR not defined, using MCUSR instead ****/
```
Comment (oliverschmidt) [inline cpu/avr/bootloader.c]: I dislike commented-out code without any hint why it is commented-out.

File: cpu/avr/watchdog.c [hunk 5]
```diff
@@ -67,7 +67,7 @@
 /* MCUSR is a deprecated name but older avr-libc versions may define it */
 #if !defined (MCUCSR)
 # if defined (MCUSR)
-#  warning *** MCUCSR not defined, using MCUSR instead ***
+/*#  warning *** MCUCSR not defined, using MCUSR instead ****/
```
Comment (oliverschmidt) [inline cpu/avr/watchdog.c]: I dislike commented-out code without any hint why it is commented-out.

File: cpu/msp430/cc2420-arch-sfd.c [hunk 6]
```diff
@@ -45,6 +45,7 @@ ISR(TIMERB1, cc2420_timerb1_interrupt)
   ENERGEST_ON(ENERGEST_TYPE_IRQ);
   /* always read TBIV to clear IFG */
   tbiv = TBIV;
+  (void)tbiv;
```
Comment (oliverschmidt) [inline cpu/msp430/cc2420-arch-sfd.c]: I suggest to either look for an alternative like defining `tbiv`as `volatile` or at least add a comment why this line was added.

File: cpu/msp430/cc2520-arch-sfd.c [hunk 7]
```diff
@@ -44,6 +44,7 @@ ISR(TIMERB1, cc2520_timerb1_interrupt)
   ENERGEST_ON(ENERGEST_TYPE_IRQ);
   /* always read TBIV to clear IFG */
   tbiv = TBIV;
+  (void)tbiv;
```
Comment (oliverschmidt) [inline cpu/msp430/cc2520-arch-sfd.c]: I suggest to either look for an alternative like defining `tbiv`as `volatile` or at least add a comment why this line was added.

File: examples/er-rest-example/resources/res-b1-sep-b2.c [hunk 8]
```diff
@@ -67,9 +67,9 @@ res_post_handler(void *request, void *response, uint8_t *buffer, uint16_t prefer
     coap_separate_accept(request, &request_metadata);
 
     /* Need Time for calculation now */
-    uint32_t i;
+    int i;
```
Comment (oliverschmidt) [inline examples/er-rest-example/resources/res-b1-sep-b2.c]: What's wrong with `i` being unsigned if it's explicitly printed as `%u` ?

File: dev/cc2520/cc2520.c [hunk 9]
```diff
@@ -217,6 +217,7 @@ flushrx(void)
   uint8_t dummy;
 
   CC2520_READ_FIFO_BYTE(dummy);
+  (void)dummy;
```
Comment (oliverschmidt) [inline dev/cc2520/cc2520.c]: I suggest to either look for an alternative like defining `dummy`as `volatile` or at least add a comment why this line was added.

File: core/ctk/ctk-filedialog.c [hunk 10]
```diff
@@ -155,14 +155,14 @@ ctk_filedialog_eventhandler(struct ctk_filedialog_state *s,
       }
       return 1; 
     } else if(ev == ctk_signal_keypress) {
-      if((ctk_arch_key_t)data == CH_CURS_UP) {
+      if((unsigned long)data == CH_CURS_UP) {
```
Comment (oliverschmidt) [inline core/ctk/ctk-filedialog.c]: Why `unsigned long`? The "original" data type is `size_t`:
```
process_post_synch(window->owner, ctk_signal_keypress, (process_data_t)(size_t)c);
```
Apart from that it isn't optimal to comapre (at least) 32 bits when comapring 8 bits is enough. The "correct" solution is present i.e. in calc.c:
```
if(ev == ctk_signal_keypress) {
  if((char)(size_t)data >= '0' &&
     (char)(size_t)data <= '9') {
```

File: core/ctk/ctk-filedialog.c (hunk 3 above)
Comment (oliverschmidt) [inline core/ctk/ctk-filedialog.c]: Why `unsigned long`? The "original" data type is `size_t`:
```
process_post_synch(window->owner, ctk_signal_keypress, (process_data_t)(size_t)c);
```
Apart from that it isn't optimal to comapre (at least) 32 bits when comapring 8 bits is enough. The "correct" solution is present i.e. in calc.c:
```
if(ev == ctk_signal_keypress) {
  if((char)(size_t)data >= '0' &&
     (char)(size_t)data <= '9') {
```

**Comment by oliverschmidt** [comment]
> Would be great if someone else could look through the changes too, as there are many of them.

Done ;-)

**Comment by simonduq** [comment]
Thanks, I'm on it!

File: cpu/msp430/cc2420-arch-sfd.c (hunk 6 above)
Comment (simonduq) [inline cpu/msp430/cc2420-arch-sfd.c]: `volatile` doesn't help, am adding a comment

File: examples/er-rest-example/resources/res-b1-sep-b2.c (hunk 8 above)
Comment (simonduq) [inline examples/er-rest-example/resources/res-b1-sep-b2.c]: The fix was on the type size rather than signedness. Now switching to unsigned, thanks

**Comment by simonduq** [comment]
I fixed all your comments except the ctk-related ones (because I'm clueless on ctk), could you please do a PR with your suggested fix on https://github.com/simonduq/contiki/tree/pr/fix-warnings ?
Thanks for the eagle eyes!

**Comment by oliverschmidt** [comment]
> could you please do a PR with your suggested fix on https://github.com/simonduq/contiki/tree/pr/fix-warnings ?

Done: https://github.com/simonduq/contiki/pull/34

> Thanks for the eagle eyes!

:-)

**Comment by simonduq** [comment]
Great, now everything is fixed here!
Let me add a :+1: to all commits in this PR that are not mine ;)

**Comment by adamdunkels** [comment]
Great comments and fixes @oliverschmidt! I'll go through this again as soon as I can!

**Comment by drugo72** [comment]
I have some compiler warning fixes for the AVR but I need to understand how to create a -Werror Travis build. @adamdunkels mentions that ```clock_time_t``` is too small on AVR platforms and need to fixed but I'm almost ready to create a PR for this (see also #1271). Where to create these two PRs, upstream or in https://github.com/simonduq/contiki/tree/pr/fix-warnings?

**Comment by simonduq** [comment]
To compile with `-Werror`, https://github.com/simonduq/contiki/commit/5b2a65758ef44ed24722e17d4a0d7395c98e458a sets a makefile flag `WERROR` for all compilation tests. https://github.com/simonduq/contiki/commit/11f4a953ebda73c1485a2c82e8366e959f294723 sets the gcc flag `-Werror` when `WERROR` is set, for cc2538 and cc26xx.

If your fixes are non-controversial simply send a PR to my repo. If they need more discussion and input from the community, a PR to upstream is probably a better choice.

**Comment by drugo72** [comment]
https://github.com/drugo72/contiki/tree/pr-fix-avr-warnings (branched from https://github.com/simonduq/contiki/tree/pr/fix-warnings) implements WERROR for the AVR cpu and fixes compiler warnings to pass the Travis build when WERROR is set.

However I'm not sending a PR to the simonduq's pr/fix-warnings branch because to pass the tests I had to temporarily cherry-pick the two commits in my PRs #1271 and #1310.
So I should wait until the PRs are merged (hopefully) and delete the temporary commits from my pr-fix-avr-warnings branch.

Do you plan to merge pr/fix-warnings soon or is this a long living branch?

**Comment by simonduq** [comment]
Thanks @drugo72. This is not intended as a long-living branch, will be merged as soon as enough people are confident about it.

**Comment by g-oikonomou** [comment]
Where are we on this one? This required a fair bit of effort from multiple people and it would be nice to consider / finalise / merge while it's still hot, lest we end up letting things slip on us again.

**Comment by simonduq** [comment]
It's good to merge as far as I'm concerned

**Pull request merged**